sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2140,21 +2140,34 @@ case class ElementAt(left: Expression, right: Expression) extends GetMapValueUti
   }
 
   override def inputTypes: Seq[AbstractDataType] = {
-    Seq(TypeCollection(ArrayType, MapType),
-      left.dataType match {
-        case _: ArrayType => IntegerType
-        case _: MapType => mapKeyType
-        case _ => AnyDataType // no match for a wrong 'left' expression type
-      }
-    )
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, hasNull), e2: IntegralType) if (e2 != LongType) =>
+        Seq(ArrayType(e1, hasNull), IntegerType)
+      case (MapType(keyType, valueType, hasNull), e2) =>
+        TypeCoercion.findTightestCommonType(keyType, e2) match {
+          case Some(dt) => Seq(MapType(dt, valueType, hasNull), dt)
+          case _ => Seq.empty
```

Comment:
Previously it will do implicitly conversion on the right expr to map's key type. Map type is not converted at all.

This change finds the common type between right expr and map's key type and will do conversion on map and right expr.

Can you provide an example showing why this is needed?

Reply:
@viirya 
```select element_at(map(1,"one", 2, "two"), 2.2D);```
In this case, there is no key with value 2.2 in the map, right ? But we return "two" as the answer. If we simply cast the right side to the key type of the map, then we will cast 2.2 down to 2 and find "two".  But by finding a common type, we will compare both sides as double and correctly return "null" as the answer.

Here is presto's output : 
```SQL
presto:default> select element_at(MAP(array[1], array[1]), 1.1);
 _col0 
-------
 NULL  
(1 row)

Query 20180921_051929_00018_iif98, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:default> select element_at(MAP(array[1], array[1]), 1.0);
 _col0 
-------
     1 
(1 row)
```

Reply:
Thanks @dilipbiswal for clarifying it. Can you update the PR description with such example, I feel it is more clear.

Reply:
@viirya Thanks .. i have updated the description. Please let me know if it looks okay to you or anything is amiss.